### PR TITLE
PLANNER-2142 Make OptaPlanner 8 a part of kogito release pipeline

### PR DESCRIPTION
**Pull request by rsynek**
### JIRA

https://issues.redhat.com/browse/PLANNER-2142

### Referenced pull requests
https://github.com/kiegroup/kogito-pipelines/pull/69
https://github.com/kiegroup/kogito-examples/pull/402

### Checklist
- [x] Documentation updated if applicable.
- [x] Upgrade recipe provided if applicable.

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add comment: <b>Jenkins run fdb</b>
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: Jenkinsfile.deploy [hunk 1]
```diff
@@ -129,6 +247,55 @@ void runMaven(String goals, String directory, boolean skipTests = false, List pr
...
+            // In case of release, we already have the settings.xml
+            maven.runMaven(mvnCmd, skipTests, ['-fae'])
+        } else {
+            maven.runMavenWithSubmarineSettings(mvnCmd, skipTests)
+        }
+    }
+}
+
+void updateKogitoVersion(String newVersion) {
+    sh "mvn versions:set-property -Dproperty=version.org.kie.kogito -DnewVersion=$newVersion -DallowSnapshots=true -DgenerateBackupPoms=false"
```
Comment (rsynek) [inline Jenkinsfile.deploy:260]: May be worth moving to a shared library (after parametrization).

File: pom.xml [hunk 2]
```diff
@@ -34,6 +34,8 @@
   </licenses>
 
   <properties>
+    <!-- TODO: remove before merging -->
+    <enforcer.skip>true</enforcer.skip>
```
Comment (rsynek) [inline pom.xml:38]: Required just for testing the pipeline. Enforcer fails the build due to the http:// protocol of the testing Nexus server (https:// required).

File: Jenkinsfile.deploy [hunk 3]
```diff
@@ -42,37 +56,118 @@ pipeline {
...
+                    }
                 }
             }
             post {
                 success {
                     script {
                         setDeployPropertyIfNeeded('git.branch', getBuildBranch())
                         setDeployPropertyIfNeeded('git.author', getGitAuthor())
+                        setDeployPropertyIfNeeded('project.version', getProjectVersion())
+                        setDeployPropertyIfNeeded('release', isRelease())
```
Comment (radtriste) [inline Jenkinsfile.deploy:71]: maybe add `KOGITO_VERSION` as well in `deployment.properties`, as you will need it in promote pipeline

File: Jenkinsfile.deploy [hunk 4]
```diff
@@ -42,37 +56,118 @@ pipeline {
...
+        stage('Setup Maven release config'){
+            when {
+                expression { return isRelease() }
+            }
+            steps {
+                echo 'Setup Maven release config'
+                configFileProvider([configFile(fileId: maven.getSubmarineSettingsXmlId(), targetLocation: 'maven-settings.xml', variable: 'MAVEN_SETTINGS_FILE')]){
+                    sh "echo '-B -s ${MAVEN_SETTINGS_FILE}' | tee optaplanner/.mvn/maven.config"
+                    // TODO: figure out why the assert does not work
+                    // assert params.MAVEN_DEPENDENCIES_REPOSITORY != '' // OptaPlanner depends on Kogito Runtimes
```
Comment (radtriste) [inline Jenkinsfile.deploy:101]: this parameter should not be mandatory. runtimes could have been released already. thus this pipeline would not need any `MAVEN_DEPENDENCIES_REPOSITORY`

Comment (radtriste) [reply Jenkinsfile.deploy:101]: just thinking I need to update that as well on examples that we should check `params.MAVEN_DEPENDENCIES_REPOSITORY` is not empty before doing the change ...

File: Jenkinsfile.deploy [hunk 5]
```diff
@@ -42,37 +56,118 @@ pipeline {
...
+            when {
+                expression { return getProjectVersion() != '' }
+            }
+            steps {
+                script {
+                    // To make sure that the project will resolve during version upgrade.
+                    mavenCleanInstall('optaplanner', true, [], '-U')
+                    dir('optaplanner') {
+                        maven.mvnVersionsSet(getProjectVersion())
+                        updateKogitoVersion(getKogitoVersion())
```
Comment (radtriste) [inline Jenkinsfile.deploy:123]: what happen if `KOGITO_VERSION` is empty ?

File: Jenkinsfile.deploy (hunk 1 above)
Comment (radtriste) [reply Jenkinsfile.deploy:260]: I do agree

File: Jenkinsfile.promote [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+boolean isRelease() {
+    return params.RELEASE || (getDeployProperty('release') == 'true')
+}
+
+String getProjectVersion() {
+    return getParamOrDeployProperty('PROJECT_VERSION', 'project.version')
+}
+
+String getKogitoVersion() {
+    return getParamOrDeployProperty('KOGITO_VERSION', 'kogito.version')
```
Comment (radtriste) [inline Jenkinsfile.promote:167]: here is the needed `kogito.version` from `deployment.properties`

File: Jenkinsfile.deploy (hunk 4 above)
Comment (radtriste) [reply Jenkinsfile.deploy:101]: already implemented in fact on examples.See https://github.com/kiegroup/kogito-examples/blob/master/Jenkinsfile.deploy#L97

File: Jenkinsfile.deploy (hunk 3 above)
Comment (rsynek) [reply Jenkinsfile.deploy:71]: I get the version there via a parameter, I suppose. But +1 to make it consistent.

File: Jenkinsfile.deploy (hunk 4 above)
Comment (rsynek) [reply Jenkinsfile.deploy:101]: Good point, will remove the assert and add a condition.

**Review by ge0ffrey — APPROVED** [review]
versions-maven-plugin isn't in the jboss-parent, so it makes sense we define it

File: pom.xml (hunk 2 above)
Comment (ge0ffrey) [reply pom.xml:38]: Yes, this must not make it to master.

Comment (ge0ffrey) [reply pom.xml:38]: Maybe you can use `-Denforcer.skip=true` in the pipeline instead?

Comment (radtriste) [reply pom.xml:38]: this would be something similar to: https://github.com/kiegroup/kogito-cloud-operator/pull/618

File: Jenkinsfile.deploy (hunk 5 above)
Comment (rsynek) [reply Jenkinsfile.deploy:125]: Already fixed by making the parameter mandatory.

File: pom.xml (hunk 2 above)
Comment (rsynek) [reply pom.xml:38]: That's an option too, thanks.

Comment (ge0ffrey) [reply pom.xml:38]: +1 for doing it in the pipeline which also adds the http:// url, instead of doing it in the pom.xml

File: Jenkinsfile.deploy [hunk 7]
```diff
@@ -99,7 +99,7 @@ pipeline {
                 script {
                     echo 'Setup Maven release config'
                     configFileProvider([configFile(fileId: maven.getSubmarineSettingsXmlId(), targetLocation: 'maven-settings.xml', variable: 'MAVEN_SETTINGS_FILE')]) {
-                        sh "echo '-B -s ${MAVEN_SETTINGS_FILE}' | tee optaplanner/.mvn/maven.config"
+                        sh "echo '\n-B -s ${MAVEN_SETTINGS_FILE}' | tee -a optaplanner/.mvn/maven.config"
```
Comment (radtriste) [inline Jenkinsfile.deploy:102]: should add also here the `sh "echo '\n-Denforcer.skip=true' | tee -a .mvn/maven.config"` command or ?

**Review by radtriste — APPROVED** [review]

**Pull request merged**